File: iosApp/iosApp/Modules/ DegenCult ✅/CultProposals ✅/CultProposalsWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3lib

enum CultProposalsWireframeDestination {
    case proposal(proposal: CultProposal, proposals: [CultProposal])
    case castVote(proposal: CultProposal, approve: Bool)
    case alert(context: AlertContext)
    case getCult
}

protocol CultProposalsWireframe {
    func present()
    func navigate(to destination: CultProposalsWireframeDestination)
}

final class DefaultCultProposalsWireframe {
    private weak var parent: UIViewController!
    private let cultProposalWireframeFactory: CultProposalWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let tokenSwapWireframeFactory: TokenSwapWireframeFactory
    private let cultService: CultService
    private let walletService: WalletService

    private weak var vc: UIViewController!

    init(
        _ parent: UIViewController,
        cultProposalWireframeFactory: CultProposalWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        tokenSwapWireframeFactory: TokenSwapWireframeFactory,
        cultService: CultService,
        walletService: WalletService
    ) {
        self.parent = parent
        self.cultProposalWireframeFactory = cultProposalWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.tokenSwapWireframeFactory = tokenSwapWireframeFactory
        self.cultService = cultService
        self.walletService = walletService
    }
}

extension DefaultCultProposalsWireframe: CultProposalsWireframe {

    func present() {
        let vc = wireUp()
        parent.show(vc, sender: self)
    }

    func navigate(to destination: CultProposalsWireframeDestination) {
        switch destination {
        case let .proposal(proposal, proposals):
            cultProposalWireframeFactory.make(
                vc,
                context: .init(proposal: proposal, proposals: proposals)
            ).present()
        case let .castVote(proposal, approve):
            confirmationWireframeFactory.makeWireframe(
                presentingIn: vc,
                context: .init(
                    type: .cultCastVote(
                        .init(cultProposal: proposal, approve: approve)
                    )
                )
            ).present()
        case let .alert(context):
            alertWireframeFactory.makeWireframe(vc, context: context).present()
        case .getCult:
            guard let presentingIn = vc.navigationController else { return }
            // TODO: Review this - we only allow CULT on ETH Mainnet atm...
            let ethNetwork = Web3Network.from(Network.Companion().ethereum(), isOn: true)
            let cultToken = Web3Token.from(
                currency: Currency.Companion().cult(),
                network: ethNetwork,
                inWallet: false,
                idx: 0
            )
            tokenSwapWireframeFactory.makeWireframe(
                presentingIn: presentingIn,
                context: .init(
                    presentationStyle: .push,
                    tokenFrom: nil,
                    tokenTo: cultToken
                )
            ).present()
        }
    }
}

private extension DefaultCultProposalsWireframe {

    func wireUp() -> UIViewController {
        let vc: CultProposalsViewController = UIStoryboard(
            .cultProposals
        ).instantiate()
        let interactor = DefaultCultProposalsInteractor(
            cultService,
            walletService: walletService
        )
        let presenter = DefaultCultProposalsPresenter(
            view: vc,
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/cultproposals/CultProposalsWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.cultproposals

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.R
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.*
import com.sonsofcrypto.web3walletcore.services.cult.CultService

class DefaultCultProposalsWireframe(
    private val parent: WeakRef<Fragment>?,
    private val cultService: CultService,
    private val walletService: WalletService,
    private val networksService: NetworksService,
): CultProposalsWireframe {

    private lateinit var fragment: WeakRef<Fragment>

    override fun present() {
        val fragment = wireUp()
        this.fragment = WeakRef(fragment)
        (parent?.get() as? NavigationFragment)?.push(fragment, animated = true)
    }

    override fun navigate(destination: CultProposalsWireframeDestination) {

    }

    private fun wireUp(): Fragment {
        val view = CultProposalsFragment()
        val interactor = DefaultCultProposalsInteractor(
            cultService, walletService
        )
        val presenter = DefaultCultProposalsPresenter(
            WeakRef(view),
            this,
            interactor
        )
        view.presenter = presenter
        return view
    }
}